Add Libc.mkfifo

tkf opened the pull request:
This PR adds a `mkfifo` wrapper `Libc.mkfifo` for Unix platforms.

Python has it in [`os` stdlib](https://docs.python.org/3/library/os.html#os.mkfifo) and it's useful sometimes.  I think it'd be nice if Julia has it in Base/stdlib too.

tkf committed:
Add Libc.mkfifo

tkf committed:
Simplify mkfifo test

tkf committed:
Run mkfifo tests only on Unix

tkf committed:
Always define Libc.mkfifo docstring

tkf committed:
Mention it in NEWS.md

vtjnash commented on base/libc.jl:
```diff
@@ -10,6 +10,9 @@ import Core.Intrinsics: bitcast
 
 export FILE, TmStruct, strftime, strptime, getpid, gethostname, free, malloc, calloc, realloc,
     errno, strerror, flush_cstdio, systemsleep, time, transcode
+if Sys.isunix()
```
I think this can be unconditional

tkf replied:
I thought to make it consistent with

https://github.com/JuliaLang/julia/blob/4fa8055225dabd6619f6e70a10c1e36cb2282cc9/base/libc.jl#L16-L18

just below.  Also, it'd make `mkfi` <kbd>TAB</kbd> completes `mkfifo` in Windows while accessing to it throws:

```julia
julia> module Demo
       export foo
       end
Main.Demo

julia> using .Demo

julia> fo  # TAB
foldl   foldr    foo      for      foreach
julia> foo
ERROR: UndefVarError: foo not defined
```

So, I thought it's better to hide it in Windows.

vtjnash replied:
It's defined though, just has no methods.

tkf replied:
Ah, yes, you are right.  But what about the consistency with `GetLastError` and `FormatMessage`?  I'm OK with unconditional `export` but I just don't see a big reason to break the pattern.

vtjnash replied:
I'm okay with those being different

tkf committed:
Merge branch 'master' into mkfifo

StefanKarpinski commented:
Not sure why this is a release blocker?

JeffBezanson commented on base/libc.jl:
```diff
@@ -400,4 +403,27 @@ Interface to the C `srand(seed)` function.
+"""
+function mkfifo end
+
+if Sys.isunix()
```
Maybe better to move this check inside the method and give a clearer "not supported on this platform" error instead of a MethodError?

tkf replied:
I actually thought it's a nice pattern to not provide the method at all so that you can use `hasmethod` etc. to check if some API is implemented in a given platform. I thought it's nicer than each caller to encode such information in their code.

JeffBezanson replied:
That's an interesting point, but I don't think `hasmethod` can be used consistently to check for platform support. For example, there could be cases where certain argument values aren't supported on a certain OS.

tkf replied:
OK. It makes sense.

tkf committed:
Define mkfifo in all platforms; a better error message than MethodError

tkf committed:
Merge branch 'master' into mkfifo

tkf commented:
@vtjnash @JeffBezanson Thanks for the review. I incorporated your comments.

tkf committed:
Add a test for Windows

StefanKarpinski commented:
Looking good... All that green! @JeffBezanson want to merge if you're happy here?

musm commented on base/libc.jl:
```diff
@@ -400,4 +400,27 @@ Interface to the C `srand(seed)` function.
+"""
+    mkfifo(path::AbstractString, [mode::Integer]) -> path
+
+Make a FIFO special file (a named pipe) at `path`.  Return `path` as-is on success.
```
Probably should also mention that this is a Unix only utiliy

musm commented:
Just needs a rebase and to change the error message to   ENOT_SUP: kernel does not support fifo

vtjnash commented on base/libc.jl:
```diff
@@ -400,4 +400,27 @@ Interface to the C `srand(seed)` function.
+        systemerror("mkfifo", ret == -1)
+        return path
+    else
+        error("not supported on this platform")
```
```suggestion
        systemerror("mkfifo", ENOTSUP)
```

tkf committed:
Update base/libc.jl

Co-authored-by: Jameson Nash <vtjnash@gmail.com>

tkf committed:
Merge branch 'master' into mkfifo

tkf committed:
Update compat

tkf committed:
Merge branch 'mkfifo' of github.com:tkf/julia into mkfifo

tkf committed:
Mention that it is supported only in Unix

tkf committed:
Update error message in test

tkf commented on base/libc.jl:
```diff
@@ -402,6 +402,31 @@ Interface to the C `srand(seed)` function.
+        systemerror("mkfifo", ret == -1)
+        return path
+    else
+        systemerror("mkfifo", ENOTSUP)
```
BTW, I noticed that `ENOTSUP` doesn't exist in my local build. I guess it's OK as long as this is defined in Windows, though.

vtjnash replied:
Looks like Linux defines EOPNOTSUP instead, and our regex isn't smart enough to pick up the value for ENOTSUP from that. But CI seems to be okay with this.

tkf replied:
It seems Windows prints `ENOTSUP` as "Unknown error":

>     Error in testset file:
>     Test Failed at C:\buildbot\worker-tabularasa\tester_win64\build\share\julia\test\file.jl:1635
>       Expression: Libc.mkfifo(joinpath(pwd(), "dummy_path"))
>         Expected: "mkfifo: Operation not supported"
>          Message: "SystemError: mkfifo: Unknown error"
>
> --- https://build.julialang.org/#/builders/65/builds/5360

So, I suggest simply doing:

https://github.com/JuliaLang/julia/blob/02961cc2b25031cb83628342f4471adbc55c9cf1/base/libc.jl#L426-L428

Is it OK?

tkf committed:
Don't use ENOTSUP; Windows does not have it?

It seems Windows prints it as "Unknown error":

    Error in testset file:
    Test Failed at C:\buildbot\worker-tabularasa\tester_win64\build\share\julia\test\file.jl:1635
      Expression: Libc.mkfifo(joinpath(pwd(), "dummy_path"))
        Expected: "mkfifo: Operation not supported"
         Message: "SystemError: mkfifo: Unknown error"

https://build.julialang.org/#/builders/65/builds/5360

This reverts commit 1ff5cd7a9e9d7b023df5b04726e78d227032f23e.

tkf committed:
Use `error("mkfifo: Operation not supported")`

MagicMuscleMan commented:
As I just needed this functionality, I was going to reinvent the wheel when I stumbled across this PR.

I do not understand why it hasn't been merged. Is there a blocker in general? Otherwise, if we prepared that for 1.10 or 1.11, would it be mergeable?

vtjnash committed:
Merge branch 'master' into mkfifo

vtjnash commented:
I just got lost because the merge me tag was removed. Putting that back now.

DilumAluthge committed:
Merge branch 'master' into mkfifo

pthariensflame commented on base/libc.jl:
```diff
@@ -437,6 +437,33 @@ function srand(seed::Integer=_make_uint64_seed())
+
+`mkfifo` is supported only in Unix platforms.
+
+!!! compat "Julia 1.8"
```
Has this been corrected?  Obviously this merge was *a bit* late for 1.8.

IanButterworth replied:
Good catch. Please PR to fix
